Task-2760 uploader: update CPC failed

vichugofsl opened the pull request:
# Description
While investigating Task 2760, I reproduced the issue that occurs when for example the Monday API key in the cfg file is incorrect. A wrong key results in a 401 response, but the existing code handled it the same way as a network error or an empty response.

To address this, I improved the class: "load/MondayHTTP.py" so that Monday HTTP server errors (e.g. 401 responses) are reported in greater detail.

Added a new RunStatus component, UpdateMonday, for each fileset to indicate whether ETL succeeds or fails in updating Monday. Added logic to treat all 4xx and 5xx responses as retryable errors. Increased the retry delay to 1.5 seconds and applied exponential back-off.

Here’s an updated MondayHTTP.py with a proper exponential-backoff retry loop that:

- Retries on 429, 500, 502, 503, and 504
- Respects the Retry-After header when present (especially for 429)
- Applies jitter to avoid thundering-herd
- Logs all 401, 400, 403 and 404 errors immediately.
- New unit test for MondayHTTP

# Task
[Bug 2760](https://dev.azure.com/keyholesoftware/FCBH%20-%20Bible%20Brain/_workitems/edit/2760): uploader: update CPC failed

# How to test it
I have used the following command and I have set a wrong Monday API key to see the following outcome with the detail of the error:
``````shell
time python3 load/DBPLoadController.py test s3://etl-development-input/ "SPNBDAP2DV"

# outcome
Database 'dbp_2025_03_06' is opened.
sendRequest: Rate limited (429). Retry-After=12 sec. Response: {"errors":["Rate limit exceeded"]}
Retryable HTTP error (status=429) on attempt 1/5: HTTP 429. Sleeping 17.8 seconds before retry…
sendRequest: Rate limited (429). Retry-After=12 sec. Response: {"errors":["Rate limit exceeded"]}
Retryable HTTP error (status=429) on attempt 2/5: HTTP 429. Sleeping 15.6 seconds before retry…
sendRequest: Rate limited (429). Retry-After=12 sec. Response: {"errors":["Rate limit exceeded"]}
Retryable HTTP error (status=429) on attempt 3/5: HTTP 429. Sleeping 17.4 seconds before retry…
sendRequest: Rate limited (429). Retry-After=12 sec. Response: {"errors":["Rate limit exceeded"]}
Retryable HTTP error (status=429) on attempt 4/5: HTTP 429. Sleeping 12.9 seconds before retry…
sendRequest: Rate limited (429). Retry-After=12 sec. Response: {"errors":["Rate limit exceeded"]}
Retryable HTTP error (status=429) on attempt 5/5: HTTP 429. Sleeping 14.9 seconds before retry…
make_monday_api_call failed after 5 attempts
********** SPNBDAP2DV-UpdateMonday Tables Update failed **********
Warnings:
WARN SPNBDAP2DV chapters missing JHN 1-9, JHN 11-21, LUK 1-9, LUK 11-24, MAT 1-9, MAT 11-28, MRK 1-9, MRK 11-16
WARN SPNBDAP2DV BiblebrainLink does not have a correct path: No zip file found
openErrorReport /home/vichugof/files/errors/Errors.out
EROR SPNBDAP2DV Error synchronizing product codes: N2SPN/BDA_JHN, N2SPN/BDA_LUK, N2SPN/BDA_MAT, N2SPN/BDA_MRK error: make_monday_api_call failed after 5 attempts
Num Errors 1
Error in RunStatus {'SPNBDAP2DV': 'ok', 'UpdateMonday': 'failed'}
``````

vichugofsl commented on load/MondayHTTP.py:
```diff
@@ -116,11 +115,25 @@ def _send_request(self, request_body: bytes) -> Optional[bytes]:
+            )
+            return None
+
+        if not resp.ok:
```
I chose to log the error instead of raising an exception, because the method that invokes _send_request already raises an error when the returned value is None.

bradflood replied:
see note elsewhere... if the response code indicates rate limit, retry using exponential backoff

vichugofsl replied:
@bradflood I have improved the implementation according to the chat Grok and we have the following changes:

Here’s an updated MondayHTTP.py with a proper exponential-backoff retry loop that:

- Logs and Retries on 429, 500, 502, 503, and 504
- Respects the Retry-After header when present (especially for 429)
- Applies jitter to avoid thundering-herd
- Logs all 401, 400, 403 and 404 errors immediately.
- New unit test for MondayHTTP

bradflood commented:
I am not sure this error is caused by an incorrect Monday API key, because there is an upload attempt just a few minutes later that succeeded -- it was able to publish to Monday. 

I suspect this is a rate limiting issue.

Please do two things:
1) add a new RunStatus component called UpdateMonday, so that we have something we can report errors against in the uploader summary area. This would be similar to the filesets. If we are unable to update Monday, mark "UpdateMonday" as failed, so we can see at a glance that it failed, without looking in the logs.
2) if Monday reports a retryable error, retry the call a few times using exponential backoff

vichugofsl commented:
> I am not sure this error is caused by an incorrect Monday API key, because there is an upload attempt just a few minutes later that succeeded -- it was able to publish to Monday.
> 
> I suspect this is a rate limiting issue.
> 
> Please do two things:
> 
>     1. add a new RunStatus component called UpdateMonday, so that we have something we can report errors against in the uploader summary area. This would be similar to the filesets. If we are unable to update Monday, mark "UpdateMonday" as failed, so we can see at a glance that it failed, without looking in the logs.
> 
>     2. if Monday reports a retryable error, retry the call a few times using exponential backoff

@bradflood I have added a new RunStatus component, UpdateMonday, for each fileset to indicate if the synchronization with Monday is successful or not. Also, I have added logic to handle all 4xx and 5xx responses as retryable errors and I have increased the retry delay to 1.5 seconds and applied exponential back-off.

I shared with you an example about the output when the synchronization with Monday fails:

`````shell
Config 'test' is loaded.
Database 'dbp_2025_03_06' is opened.
sendRequest: Unexpected status code 404.
sendRequest: Unexpected status code 404.
sendRequest: Unexpected status code 404.
sendRequest: Unexpected status code 404.
sendRequest: Unexpected status code 404.
********** SPNBDAP2DV-UpdateMonday Tables Update failed **********
Warnings:
WARN SPNBDAP2DV chapters missing JHN 1-9, JHN 11-21, LUK 1-9, LUK 11-24, MAT 1-9, MAT 11-28, MRK 1-9, MRK 11-16
WARN SPNBDAP2DV BiblebrainLink does not have a correct path: No zip file found
openErrorReport /home/vichugof/files/errors/Errors.out
EROR SPNBDAP2DV Error synchronizing product codes: N2SPN/BDA_JHN, N2SPN/BDA_LUK, N2SPN/BDA_MAT, N2SPN/BDA_MRK error: make_monday_api_call failed after 5 attempts
Num Errors 1
Error in RunStatus {'SPNBDAP2DV': 'ok', 'SPNBDAP2DV-UpdateMonday': 'failed'}

``````

bradflood reviewed (changes requested):
We only want one RunStatus, called UpdateMonday. We don't need one for each fileset, and we don't want to prepend the fileset id to "UpdateMonday".
Also, we should not add UpdateMonday for text loading, just audio and video

vichugofsl commented:
> We only want one RunStatus, called UpdateMonday. We don't need one for each fileset, and we don't want to prepend the fileset id to "UpdateMonday". Also, we should not add UpdateMonday for text loading, just audio and video

@bradflood I have fixed it and I created an only one RunStatus called UpdateMonday as you can see in the following output. It will only execute for audio and video fileset as the method: synchronizeMonday has the constraint to validate the typeCode.

``````shell
EROR SPNBDAP2DV Error synchronizing product codes: N2SPN/BDA_JHN, N2SPN/BDA_LUK, N2SPN/BDA_MAT, N2SPN/BDA_MRK error: make_monday_api_call failed after 5 attempts
Num Errors 1
Error in RunStatus {'SPNBDAP2DV': 'ok', 'UpdateMonday': 'failed'}
``````

vichugofsl commented on load/MondayHTTP.py:
```diff
@@ -100,27 +120,64 @@ def _send_request(self, request_body: bytes) -> Optional[bytes]:
+            raise HTTPRetryableError(0, message=str(e))
+
+        # Handle specific status codes
+        if resp.status_code == 401:
```
However, I have created a specific log with the details of the 401 error
